### PR TITLE
Intern place projection

### PR DESCRIPTION
**Pull request by spastorino**
This should sit on top of https://github.com/rust-lang/rust/pull/65197. After that one merged, I'm gonna rebase on top of it.

The important commits are the last three and there's a bunch of code repetition that I'm going to remove but for that I need to refactor some things that probably need to be added before this PR.

Anyway this work helps as is because we can run perf tests :).

r? @oli-obk /cc @nikomatsakis

### CODE REVIEW

File: src/librustc/mir/mod.rs [hunk 1]
```diff
@@ -1857,41 +1858,44 @@ impl<'tcx> Place<'tcx> {
     pub fn return_place() -> Place<'tcx> {
         Place {
             base: PlaceBase::Local(RETURN_PLACE),
-            projection: Box::new([]),
+            projection: List::empty(),
```
Comment (oli-obk) [inline src/librustc/mir/mod.rs]: it should be possible to use some feature gates to make `List::empty` `const fn`, thus allowing  you to resolve the FIXME above.

File: src/librustc/mir/mod.rs [hunk 2]
```diff
@@ -1857,41 +1858,44 @@ impl<'tcx> Place<'tcx> {
...
         }
     }
 
-    pub fn field(self, f: Field, ty: Ty<'tcx>) -> Place<'tcx> {
-        self.elem(ProjectionElem::Field(f, ty))
+    pub fn field(self, f: Field, ty: Ty<'tcx>, tcx: TyCtxt<'tcx>) -> Place<'tcx> {
```
Comment (oli-obk) [inline src/librustc/mir/mod.rs]: all of these could become methods on `TyCtxt` similar to all the `mk_*` methods

File: src/librustc/mir/mod.rs [hunk 3]
```diff
@@ -1908,24 +1912,24 @@ impl<'tcx> Place<'tcx> {
     //
     // FIXME: can we safely swap the semantics of `fn base_local` below in here instead?
     pub fn local_or_deref_local(&self) -> Option<Local> {
-        match self {
-            Place {
-                base: PlaceBase::Local(local),
-                projection: box [],
+        match self.as_ref() {
```
Comment (oli-obk) [inline src/librustc/mir/mod.rs]: most of these could be pulled out into a separate PR, making the diff here smaller

File: src/librustc/mir/mod.rs [hunk 4]
```diff
@@ -3184,6 +3188,17 @@ impl<'tcx> TypeFoldable<'tcx> for PlaceBase<'tcx> {
     }
 }
 
+impl<'tcx> TypeFoldable<'tcx> for &'tcx ty::List<PlaceElem<'tcx>> {
```
Comment (oli-obk) [inline src/librustc/mir/mod.rs]: side-note: maybe some of the `MutVisitor`s should be `TypeFolder`s instead?

File: src/librustc/ty/codec.rs [hunk 5]
```diff
@@ -215,6 +215,20 @@ where
...
 }
 
+#[inline]
+pub fn decode_place<D>(decoder: &mut D) -> Result<mir::Place<'tcx>, D::Error>
+where
+    D: TyDecoder<'tcx>,
+{
+    let base: mir::PlaceBase<'tcx> = Decodable::decode(decoder)?;
+    let len = decoder.read_usize()?;
+    let interned: Vec<mir::PlaceElem<'tcx>> =
```
Comment (oli-obk) [inline src/librustc/ty/codec.rs]: are there no nice wrappers out there for (de)serializign a `List`?

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/librustc/mir/mod.rs (hunk 1 above)
Comment (spastorino) [reply src/librustc/mir/mod.rs]: Can't make `List::empty` a const fn. there's an `assert!(mem::align_of::<T>() <= 64);` so I get ...

```
error[E0019]: constant function contains unimplemented expression type
   --> src/librustc/ty/mod.rs:734:9
    |
734 |           assert!(mem::align_of::<T>() <= 64);
    |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ in this macro invocation
    | 
   ::: <::core::macros::builtin::assert macros>:1:1
    |
1   | / ($ cond : expr) => ({ }) ; ($ cond : expr,) => ({ }) ;
2   | | ($ cond : expr, $ ($ arg : tt) +) => ({ })
    | |__________________________________________- in this expansion of `assert!`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0019`.
error: could not compile `rustc`.
```

Comment (spastorino) [reply src/librustc/mir/mod.rs]: And I'd also get ...

```
error[E0013]: constant functions cannot refer to statics, use a constant instead
   --> src/librustc/ty/mod.rs:735:16
    |
735 |             &*(&EMPTY_SLICE as *const _ as *const List<T>)
    |                ^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0013`.
```

Comment (spastorino) [reply src/librustc/mir/mod.rs]: The EMPTY_SLICE could be a const fn too

File: src/librustc/mir/mod.rs (hunk 3 above)
Comment (spastorino) [reply src/librustc/mir/mod.rs]: Done

File: src/librustc/mir/mod.rs (hunk 4 above)
Comment (spastorino) [reply src/librustc/mir/mod.rs]: Have you seen any in particular that you consider?

File: src/librustc/ty/codec.rs (hunk 5 above)
Comment (spastorino) [reply src/librustc/ty/codec.rs]: I've fixed the deserialization and made it look a bit better. But what we have is very similar to what's done with ty https://github.com/rust-lang/rust/blob/e0fe4846e5ab3f5d8810280bbe0d217cafd491c9/src/librustc/ty/codec.rs#L247

File: src/librustc/mir/mod.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc/mir/mod.rs]: you can also just make `EMPTY_SLICE` a `const` instead of a `static`

The assert can be replaced by `[()][(mem::align_of::<T>() > 64) as usize];` (and a FIXME stating that we should use the assert again once const eval supports that)

File: src/librustc/mir/mod.rs (hunk 4 above)
Comment (oli-obk) [reply src/librustc/mir/mod.rs]: all those that just apply an operation to all types.

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 6]
```diff
@@ -273,12 +273,12 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
-                        if let Place {
-                            base: PlaceBase::Local(borrowed_local),
-                            projection: box [],
-                        } = place {
-                             if body.local_decls[*borrowed_local].name.is_some()
-                                && local != *borrowed_local
+                        if let PlaceRef {
+                            base: &PlaceBase::Local(borrowed_local),
+                            projection: &[],
+                        } = place.as_ref() {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: this could use the `as_local` method

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 7]
```diff
@@ -494,22 +494,22 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
-                            }) => {
-                                let local_decl = &self.body.local_decls[*l];
-                                if local_decl.name.is_none() {
-                                    local_decl.source_info.span
+                            Operand::Copy(place) |
+                            Operand::Move(place) => {
+                                if let PlaceRef {
+                                    base: &PlaceBase::Local(l),
+                                    projection: &[],
+                                } = place.as_ref() {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: this could use the `as_local` method

File: src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs [hunk 8]
```diff
@@ -582,17 +586,16 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
-                            | Operand::Move(Place {
-                                base: PlaceBase::Local(from),
-                                projection: box [],
-                            })
-                                if *from == target =>
-                            {
-                                target = into;
+                            Operand::Copy(place)
+                            | Operand::Move(place) => {
+                                if let PlaceRef {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs]: use as_local

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 9]
```diff
@@ -1124,10 +1123,10 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
         };
 
         let (place_desc, note) = if let Some(place_desc) = opt_place_desc {
-            let local_kind = match borrow.borrowed_place {
-                Place {
-                    base: PlaceBase::Local(local),
-                    projection: box [],
+            let local_kind = match borrow.borrowed_place.as_ref() {
```
Comment (oli-obk) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: use `as_local` here

**Review by oli-obk — CHANGES_REQUESTED** [review]
r=me with one more `as_local` call
